shell/platform/darwin/macos/framework/Source/FlutterDisplayLink.mm

```diff
@@ -0,0 +1,355 @@
+                                       CVOptionFlags* flags_out) {
+  // Hold the mutex only while copying clients.
+  std::vector<_FlutterDisplayLink*> clients;
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    for (ScreenEntry& entry : entries_) {
+      if (entry.display_link_locked == display_link) {
+        clients = entry.clients;
+        break;
+      }
+    }
+  }
+
+  CFTimeInterval timestamp = (CFTimeInterval)in_now->hostTime / CVGetHostClockFrequency();
+  CFTimeInterval target_timestamp =
+      (CFTimeInterval)in_output_time->hostTime / CVGetHostClockFrequency();
+
+  for (_FlutterDisplayLink* client : clients) {
+    [client didFireWithTimestamp:timestamp targetTimestamp:target_timestamp];
+  }
+}
+}  // namespace
+
+@interface _FlutterDisplayLinkView : NSView {
+}
+
+@end
+
+static NSString* const kFlutterDisplayLinkViewDidMoveToWindow =
+    @"FlutterDisplayLinkViewDidMoveToWindow";
+
+@implementation _FlutterDisplayLinkView
+
+- (void)viewDidMoveToWindow {
+  [super viewDidMoveToWindow];
+  [[NSNotificationCenter defaultCenter] postNotificationName:kFlutterDisplayLinkViewDidMoveToWindow
+                                                      object:self];
+}
+
+@end
+
+@implementation _FlutterDisplayLink
+
+@synthesize delegate = _delegate;
+
+- (instancetype)initWithView:(NSView*)view {
+  FML_DCHECK([NSThread isMainThread]);
+  if (self = [super init]) {
+    self->_view = [[_FlutterDisplayLinkView alloc] initWithFrame:CGRectZero];
+    [view addSubview:self->_view];
+    _paused = YES;
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(viewDidChangeWindow:)
+                                                 name:kFlutterDisplayLinkViewDidMoveToWindow
+                                               object:self->_view];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(windowDidChangeScreen:)
+                                                 name:NSWindowDidChangeScreenNotification
+                                               object:nil];
+    [self updateScreen];
+  }
+  return self;
+}
+
+- (void)invalidate {
+  @synchronized(self) {
+    FML_DCHECK([NSThread isMainThread]);
+    [_view removeFromSuperview];
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+    _view = nil;
+    _delegate = nil;
+  }
+  DisplayLinkManager::Instance().UnregisterDisplayLink(self);
+}
+
+- (void)updateScreen {
+  DisplayLinkManager::Instance().UnregisterDisplayLink(self);
+  std::optional<CGDirectDisplayID> displayId;
+  @synchronized(self) {
+    NSScreen* screen = _view.window.screen;
+    if (screen != nil) {
+      // https://developer.apple.com/documentation/appkit/nsscreen/1388360-devicedescription?language=objc
+      _display_id = (CGDirectDisplayID)[
+          [[screen deviceDescription] objectForKey:@"NSScreenNumber"] unsignedIntValue];
```

Comment:
Sure would have been nice if they'd provided a const NSString for this key! Thanks for the doc link to explain it.